### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by mickaelistria**
Refactor remote repository access to we get finer grain report about
which repos we're fetching index for, and results from one repo even if
others are pending for index download.

Signed-off-by: Mickael Istria <mistria@redhat.com>

### CODE REVIEW

File: org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenRepositoryCache.java [hunk 1]
```diff
@@ -17,71 +17,6 @@
 import org.eclipse.aether.repository.RemoteRepository;
 
 public class MavenRepositoryCache {
```
Comment (AObuchow) [inline org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenRepositoryCache.java]: Should we just delete this class as it's logic has moved to `RemoteRepositoryIndexSearcher`?

File: org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenCompletionParticipant.java [hunk 2]
```diff
@@ -245,44 +241,45 @@ private void collectionRemoteGAVCompletion(ICompletionRequest request, ICompleti
...
+								if (artifactToSearch.getGroupId().equals(VersionValidator.placeholderArtifact.getGroupId())) {
+									// Don't do a remote groupId search if the user hasn't inputed anything as there
+									// will be too many results
+									return;
+								}
+								indexSearcher.getGroupIds(artifactToSearch, index)
+										.stream().map(groupId -> toCompletionItem(groupId, "GroupId", range)).forEach(items::add);
+								return;
+							case "artifactId":
+								indexSearcher.getArtifactIds(artifactToSearch, index).stream().map(artifactInfo -> toCompletionItem(artifactInfo.getArtifactId(), artifactInfo.getArtifactId(), range)).forEach(items::add);
```
[the remark this answers is not in the document]
Comment (AObuchow) [reply org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenCompletionParticipant.java]: It was originally done [like this](https://github.com/AObuchow/lsp4xml-extensions-maven/blob/master/org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/searcher/RemoteRepositoryIndexSearcher.java#L184)

[the remark this answers is not in the document]
Comment (mickaelistria) [reply org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenCompletionParticipant.java]: you're right. my mistake!

File: org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenRepositoryCache.java (hunk 1 above)
Comment (mickaelistria) [reply org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenRepositoryCache.java]: Sure.

**Comment by mickaelistria** [comment]
Thanks Andrew. I just pushed a new commit that should include your suggestions.

**Comment by AObuchow** [comment]
I tried running the tests on my machine and they're failing (due to overlapping file lock exception I believe) - I think this is acceptable as we can't properly fix this until we upgrade to lsp4xml 0.9.2-SNAPSHOT or LemMinX when it's on eclipse.org. 

Manual testing works, although I noticed duplicate artifactId's
image
My guess is that two maven central indexes are being added to the list of remote repositories, which is something I had run into myself. I had solved it by checking for duplicate remote repository Id's (but this solution could lead to other bugs). From what I understand, you're using a hash to ensure duplicate repo's don't cause any file conflicts with eachother. Alternatively, we could filter for duplicate artifactId's (maybe prioritizing the `artifactInfo` that's `getDescription()` is non null).

Additionally, storing the maven indexes/caches in `~/.m2/repository/` might not be the best place, as it causes the indexes to appear in the local groupId completion. 
image
Perhaps we should store the indexes somewhere like `~/.m2/index/`?

Despite these observations, I'm fine with merging this commit and addressing those issues in seperate bugs/patches. @mickaelistria please let me know if you would like to try fixing those issues in this patch, or if we should leave it for a future fix (I am comfortable with fixing them myself)

**Comment by mickaelistria** [comment]
If it's OK with you, then let's merge now and track further issues separately.

**Comment by AObuchow** [comment]
> If it's OK with you, then let's merge now and track further issues separately.
That's fine with me, going to merge.

**Pull request merged**